desc/objectives/_neoclassical.py

```diff
@@ -0,0 +1,263 @@
+    Evaluation of 1/ν neoclassical transport in stellarators.
+    V. V. Nemov, S. V. Kasilov, W. Kernbichler, M. F. Heyn.
+    Phys. Plasmas 1 December 1999; 6 (12): 4622–4632.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium that will be optimized to satisfy the Objective.
+    target : {float, ndarray, callable}, optional
+        Target value(s) of the objective. Only used if bounds is None.
+        Must be broadcastable to Objective.dim_f. If a callable, should take a
+        single argument ``rho`` and return the desired value of the profile at those
+        locations. Defaults to 0.
+    bounds : tuple of {float, ndarray, callable}, optional
+        Lower and upper bounds on the objective. Overrides target.
+        Both bounds must be broadcastable to Objective.dim_f.
+        If a callable, each should take a single argument ``rho`` and return the
+        desired bound (lower or upper) of the profile at those locations.
+    weight : {float, ndarray}, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        Must be broadcastable to Objective.dim_f
+    normalize : bool, optional
+        Whether to compute the error in physical units or non-dimensionalize.
+    normalize_target : bool, optional
+        Whether target and bounds should be normalized before comparing to computed
+        values. If `normalize` is ``True`` and the target is in physical units,
+        this should also be set to True.
+    loss_function : {None, 'mean', 'min', 'max'}, optional
+        Loss function to apply to the objective values once computed. This loss function
+        is called on the raw compute value, before any shifting, scaling, or
+        normalization.
+    deriv_mode : {"auto", "fwd", "rev"}
+        Specify how to compute Jacobian matrix, either forward mode or reverse mode AD.
+        "auto" selects forward or reverse mode based on the size of the input and output
+        of the objective. Has no effect on self.grad or self.hess which always use
+        reverse mode and forward over reverse mode respectively.
+    rho : ndarray
+        Unique coordinate values specifying flux surfaces to compute on.
+    alpha : ndarray
+        Unique coordinate values specifying field line labels to compute on.
+    knots_per_transit : int
+        Number of points per toroidal transit at which to sample data along field
+        line. Default is 100.
+    num_transit : int
+        Number of toroidal transits to follow field line.
+        For axisymmetric devices, one poloidal transit is sufficient. Otherwise,
+        more transits will give more accurate result, with diminishing returns.
+    num_quad : int
+        Resolution for quadrature of bounce integrals. Default is 32.
+    num_pitch : int
+        Resolution for quadrature over velocity coordinate. Default 50.
+    batch : bool
+        Whether to vectorize part of the computation. Default is true.
+    num_well : int
```

Comment:
How much does this effect performance? Is there some heuristic we can come up with as a default?

I feel like something along the lines of `(M + abs(iota)*N)*num_transit` would be reasonable? (maybe with some fudge factors?)

Do you have an estimate of how many wells there are from the example cases you've run?

Reply:
Usually you can set it to `knots_per_transit` which reduces by factor of 3 without affecting anything, because if `knots_per_transit` is low enough that a single cubic spline recreates an entire well, then the reconstruction of |B| doesn't have sufficient resolution

Reply:
> How much does this effect performance?
I've saved a couple gigabytes of memory by setting to `knots_per_transit // 3`

Reply:
> (M + abs(iota)*N)*num_transit
ok we could try that

Reply:
I am having trouble testing whether this is worth it.

On my machine (cpu) running `pytest -k test_effective_ripple`: regardless of whether compute one pitch at a time (using below code), or `batch=False`, or hardcoding `num_well=10` (this should miss almost all wells), or all three, I do not see change in peak memory usage. On the other hand, increasing the parameters that control the transforms size, e.g. `knots_per_transit`,  `num_rho`, or `num_alpha` signifcantly increases memory use.

I think since the transforms are the dominant memory cost this is expected. I do remember running an optimization with gpu on the cluster where `num_well` did have an effect. Maybe the memory penalty of using AD increases the cost of `num_well` to something that is measurable in that case.

This issue with using the maximum frequency heuristic above is that it depends on an optimizable quantity, which means the compute function for `effective_ripple` will need to be recompiled each time as `num_well` is a static argument.

```python
# bounce integral stuff one pitch at a time
    def eps_32(data):
        """(∂ψ/∂ρ)⁻² B₀⁻² ∫ dλ λ⁻² ∑ⱼ Hⱼ²/Iⱼ."""
        bounce = Bounce1D(grid, data, quad, automorphism=None, is_reshaped=True)

        def _eps_32(pitch_inv):
            points = bounce.points(pitch_inv, num_well=num_well)
            # batch=false means one bounce integral at a time
            H = bounce.integrate(
                dH,
                pitch_inv,
                data["|grad(rho)|*kappa_g"],
                points=points,
                batch=batch,
            )
            I = bounce.integrate(dI, pitch_inv, points=points, batch=batch)
            return safediv(H**2, I).sum()

        return (
            imap(_eps_32, data["pitch_inv"])  # computes one pitch at atime
            * data["pitch_inv"] ** (-3)
            * data["pitch_inv weight"]
        ).sum(axis=-1)
```

Reply:
Could do `(M + N)*num_transit`

Reply:
yeah, if we only need a rough upper bound we can assume iota~1. And maybe throw a factor of 4-5 out front to account for even/odd frequencies and nonlinearities due to lambda?

Reply:
> throw a factor of 4-5... nonlinearities due to lambda

In theory $A + B  \sim NFP (M + N)$. A good heuristic for the max frequency would be $eq.NFP * (eq.M + eq.N)$. I think the max frequency heuristic is too loose. I think it's better to just explain this as is done in #1290 and let user pick.